Network.lua

```diff
@@ -4,98 +4,164 @@ require 'BRNN'
-    if self.gpu then
-        ctcCriterion = nn.CTCCriterion():cuda()
+    local sizes = torch.Tensor()
+    if self.nGPU > 0 then
+        ctcCriterion = ctcCriterion:cuda()
         inputs = inputs:cuda()
+        sizes = sizes:cuda()
     end
 
+    -- def loading buf and loader
+    local loader = loader(self.lmdb_path)
+    local spect_buf, label_buf, sizes_buf
+
+    -- load first batch
+    local inds = self.indexer:nxt_inds() -- use nxt_inds if zero-mask is done
+    self.pool:addjob(function()
+                        return loader:nxt_batch(inds, false)
+                    end,
+                    function(spect,label,sizes)
+                        spect_buf=spect
+                        label_buf=label
+                        sizes_buf=sizes
+                    end
+                    )
+
+    -- ===========================================================
+    -- define the feval
+    -- ===========================================================
     local function feval(x_new)
-        local inputsCPU, targets = dataset:nextData()
-        -- transfer over to GPU
-        inputs:resize(inputsCPU:size()):copy(inputsCPU)
+        --------------------- data load ------------------------
+        self.pool:synchronize()                         -- wait previous loading
+        local inputsCPU,sizes,targets = spect_buf,sizes_buf,label_buf   -- move buf to training data
+        inds = self.indexer:nxt_inds()                  -- load nxt batch
+        self.pool:addjob(function()
+                            return loader:nxt_batch(inds, false)
+                        end,
+                        function(spect,label,sizes)
+                            spect_buf=spect
+                            label_buf=label
+                            sizes_buf=sizes
+                        end
+                        )
+        --------------------- fwd and bwd ---------------------
+        inputs:resize(inputsCPU:size()):copy(inputsCPU) -- transfer over to GPU
+        sizes = self.calSize(sizes)
         gradParameters:zero()
-        local predictions = self.model:forward(inputs)
-        local loss = ctcCriterion:forward(predictions, targets)
+        cutorch.synchronize()
+        local predictions = self.model:forward({inputs, sizes})
+        local loss = ctcCriterion:forward(predictions, targets, sizes)
         self.model:zeroGradParameters()
         local gradOutput = ctcCriterion:backward(predictions, targets)
         self.model:backward(inputs, gradOutput)
+        cutorch.synchronize()
         return loss, gradParameters
     end
 
+    -- ==========================================================
+    -- training
+    -- ==========================================================
     local currentLoss
     local startTime = os.time()
-    local dataSetSize = dataset:size()
+    -- local dataSetSize = self.indexer.len_num -- obtained when calling prep_same_len_inds
+    local dataSetSize = 48
```

Comment:
Would the below work? This involved me having to create a batch_size variable in the init using the networkParams.

``` local dataSetSize = math.ceil(self.indexer.lmdb_size / self.batch_size)```